docs/book/src/reference/watching-resources/testdata/external-indexed-field/controller.go

```diff
@@ -0,0 +1,210 @@
+limitations under the License.
+*/
+// +kubebuilder:docs-gen:collapse=Apache License
+
+/*
+Along with the standard imports, we need additional controller-runtime and apimachinery libraries.
+*/
+package external_indexed_field
+
+import (
+	"context"
+	"github.com/prometheus/common/log"
+
+	"github.com/go-logr/logr"
+	kapps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	appsv1 "tutorial.kubebuilder.io/project/api/v1"
+)
+
+/*
+Determine the path of the field in the ConfigDeployment CRD that we wish to use as the "object reference".
+This will be used in both the indexing and watching.
+*/
+const (
+	configMapField = ".spec.configMap"
+)
+
+/*
+*/
+
+// ConfigDeploymentReconciler reconciles a ConfigDeployment object
+type ConfigDeploymentReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+// +kubebuilder:docs-gen:collapse=Reconciler Declaration
+
+/*
+There are two additional resources that the controller needs to have access to, other than ConfigDeployments.
+- It needs to be able to fully manage Deployments, as well as check their status.
+- It also needs to be able to get, list and watch ConfigMaps.
+All 3 of these are important, and you will see usages of each below.
+*/
+
+// +kubebuilder:rbac:groups=apps.tutorial.kubebuilder.io,resources=configdeployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.tutorial.kubebuilder.io,resources=configdeployments/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+/*
+`Reconcile` will be in charge of reconciling the state of ConfigDeployments.
+ConfigDeployments are used to manage Deployments whose pods are updated whenever the configMap that they use is updated.
+
+For that reason we need to add an annotation to the PodTemplate within the Deployment we create.
+This annotation will keep track of the latest version of the data within the referenced ConfigMap.
+Therefore when the version of the configMap is changed, the PodTemplate in the Deployment will change.
+This will cause a rolling upgrade of all Pods managed by the Deployment.
+
+Skip down to the `SetupWithManager` function to see how we ensure that `Reconcile` is called when the referenced `ConfigMaps` are updated.
+*/
+func (r *ConfigDeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
```

Comment:
You are using kubebuilder v2.3.1 or lower. Current documentation is already targeting v3.0.0 which defaults `go/v3` plugin. Could you please update the example to use this plugin? I would suggest you create a dummy project with a single API and its controller with your current version of kubebuilder, and then use the latest v3.0.0-alphaX release to create an equivalent project and compare the `*_types.go` and `*_controller.go` files. I know that some of the changes are:
- The [`Reconcile` function](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.7.0/pkg/reconcile#Func) signature has changed to also receive a `context.Context` as the first parameter. This removes the need to call `context.Background()` (which by the way you were assigning to `_` but later using `ctx`).
- RBAC for finalizers subresources was added.
- Both functions were added some comments to document them.

I am listing them from memory and without knowing exactly which version of kubebuilder you are using, so double check this please.

Reply:
Yep, using Kubebuilder v2.3.1.

I can do that for sure.